feat(code-snippet): Add new Code snippet variation and styles

alisonjoseph opened the pull request:
## Overview

Ref https://github.ibm.com/carbon/issues/issues/658 
https://github.com/carbon-design-system/carbon-components/issues/535

<img width="218" alt="screen shot 2018-05-10 at 1 42 43 pm" src="https://user-images.githubusercontent.com/2753488/39887803-bd36e2d8-5458-11e8-94ce-2ad28aeb992e.png">
<img width="241" alt="screen shot 2018-05-10 at 1 42 37 pm" src="https://user-images.githubusercontent.com/2753488/39887805-bd74141e-5458-11e8-92be-bd72c8ff2ae1.png">
<img width="636" alt="screen shot 2018-05-10 at 1 42 26 pm" src="https://user-images.githubusercontent.com/2753488/39887806-bd859342-5458-11e8-899e-874e783b62b4.png">
<img width="628" alt="screen shot 2018-05-10 at 1 42 19 pm" src="https://user-images.githubusercontent.com/2753488/39887807-bd96678a-5458-11e8-95d3-115bbed3d9de.png">
<img width="817" alt="screen shot 2018-05-10 at 1 42 04 pm" src="https://user-images.githubusercontent.com/2753488/39887808-bdb209cc-5458-11e8-8d0b-e7612ed38391.png">


### Added

Inline code snippet (and light variation)
Migration doc to 9.x

### Changed

Renamed terminal to single and code to multi (noted in migration docs)

## Testing / Reviewing

http://localhost:3000/demo/code-snippet

alisonjoseph committed:
feat(code-snippet): Update code snippet styles and names

alisonjoseph committed:
feat(code-snippet): Update code snippet styles

alisonjoseph committed:
feat(code-snippet): Update font size

alisonjoseph committed:
chore: cleanup html files

alisonjoseph committed:
feat(code-snippet): Update feedback copy tooltip to match new styles

asudoh commented on src/components/code-snippet/code-snippet.js:
```diff
@@ -0,0 +1,68 @@
+   */
+  static options = {
+    selectorInit: '.bx--snippet--multi .bx--snippet-btn--expand',
+    showMore: 'Show more',
```
Great to embed those translatable strings in data attribute so the consumer’s code does not have to deal with our vanilla JavaScript code for translation - e.g. `this.element.getAttribute(this.options.attribShowMoreText)` where `attribShowMoreText` is `data-show-more-text`.

asudoh commented on src/components/code-snippet/code-snippet.js:
```diff
@@ -0,0 +1,68 @@
+    this.element.parentElement.classList.toggle('bx--snippet--expand');
+
+    if (this.element.parentElement.classList.contains('bx--snippet--expand')) {
+      this.element.querySelectorAll('.bx--snippet-btn--text')[0].innerHTML = this.options.showLess;
```
`innerHTML` runs a HTML parser - You may want to try `.textContent`.

asudoh commented on src/components/code-snippet/code-snippet.js:
```diff
@@ -0,0 +1,68 @@
+
+  _initCodeSnippet() {
+    this.element.querySelectorAll('.bx--snippet-btn--text')[0].innerHTML = this.options.showMore;
+    if (this.element.parentElement.offsetHeight < 288) {
```
Great to do `this.element.parentElement.offsetHeight < this.options.summaryHeight` where `summaryHeight` is `288`.

asudoh commented on src/components/code-snippet/multi-line-code-snippet.html:
```diff
@@ -0,0 +1,40 @@
+<div class="bx--snippet bx--snippet--multi">
```
Great to add `data-code-snippet` here (and change `selectorInit` option accordingly) to make our JavaScript doesn’t have to deal with changes in CSS class structure. This means the root element for our JavaScript code is the whole snippet instead of the expando. Doing so allows us to remove an assumption on the DOM structure - The expando being a direct child of the code snippet.

asudoh commented on src/components/code-snippet/code-snippet.js:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  _handleClick() {
+    this.element.parentElement.classList.toggle('bx--snippet--expand');
```
Great to define `classExpanded` option (value is `bx--snippet--expand`) for more flexible custom theming, etc. Same for all classes/selectors.

asudoh commented on src/components/code-snippet/inline-code-snippet-light.html:
```diff
@@ -0,0 +1,7 @@
+
```
Do we really want the whitespace here…?

asudoh reviewed (commented):
Great to see you implemented the expando quickly @alisonjoseph!

alisonjoseph committed:
chore: cleanup javascript

alisonjoseph commented:
@asudoh I think I've addressed most of your concerns and updated the PR. What is the best way to target the button only for the handleClick (and not the entire element)?

tw15egan commented on src/components/copy-button/_copy-button.scss:
```diff
@@ -23,25 +23,30 @@
 
     &:after {
       top: 0.85rem;
-      width: 0.5rem;
```
I think this just needs a `left: -.3px;` to center the top arrow 

<img width="247" alt="screen shot 2018-05-11 at 1 46 29 pm" src="https://user-images.githubusercontent.com/11928039/39942006-d99b7d78-5523-11e8-94ac-18b450117764.png">

<img width="650" alt="screen shot 2018-05-11 at 1 54 52 pm" src="https://user-images.githubusercontent.com/11928039/39942011-de3f18bc-5523-11e8-96e3-6b594491e02c.png">

tw15egan reviewed (changes requested):
Other than a small change, looks great!

alisonjoseph committed:
feat(code-snippet): Update copy tooltip styles

asudoh commented on src/components/code-snippet/code-snippet.js:
```diff
@@ -0,0 +1,77 @@
+   */
+  static options = {
+    selectorInit: '[data-code-snippet]',
+    showMoreText: 'data-show-more-text',
```
Let’s rename it to `attribShowMoreText`/`attribShowLessText` - Other components take this naming rule.

asudoh commented on src/components/code-snippet/code-snippet.js:
```diff
@@ -0,0 +1,77 @@
+    const expandBtn = this.element.querySelector(this.options.classExpandText);
+    this.element.classList.toggle(this.options.classExpanded);
+
+    if (this.element.classList.contains(this.options.classExpanded)) {
```
Let’s put a guard against NPE (when `expandBtn` does not exist), by checking the existence here.

asudoh replied:
For better explanation - Here's where the idea came from: https://github.com/carbon-design-system/carbon-components/pull/12

alisonjoseph replied:
👍

asudoh commented on src/components/code-snippet/code-snippet.js:
```diff
@@ -0,0 +1,77 @@
+  _initCodeSnippet() {
+    const expandBtn = this.element.querySelector(this.options.classExpandText);
+
+    expandBtn.textContent = expandBtn.getAttribute(this.options.showMoreText);
```
Let’s put a guard against NPE (when `expandBtn` does not exist), by checking the existence here.

asudoh replied:
For better explanation - Here's where the idea came from: https://github.com/carbon-design-system/carbon-components/pull/12

asudoh reviewed (commented):
> What is the best way to target the button only for the `handleClick` (and not the entire element)?

@alisonjoseph Thank you for asking! There are two ways to make the event handler only for the button:

1. Adding `if (eventMatches(evt, this.options.selectorExpandBtn))` check where `this.options.selectorExpandBtn` is `.bx--snippet-btn--expand`
2. Attach an event handler to the button instead of to the root element

Please don't hesitate to ping me if you have any further questions - Thanks!

alisonjoseph committed:
chore: cleanup js

alisonjoseph committed:
feat(code-snippet): Update click event to only happen on button

alisonjoseph committed:
chore: Add back support for deprecated class names

marijohannessen commented:
![screen shot 2018-05-14 at 11 31 52 am](https://user-images.githubusercontent.com/5447411/40010616-e4964476-576a-11e8-85a9-920cd63c208c.png)
![screen shot 2018-05-14 at 11 31 45 am](https://user-images.githubusercontent.com/5447411/40010617-e4b4d88c-576a-11e8-97de-217b874f5c4b.png)

Are these supposed to be a bit off-centered?

alisonjoseph commented:
@marijohannessen I'm going to review the placement of the tooltip with @tay-aitken this afternoon. I agree it looks a little off.

alisonjoseph committed:
feat(code-snippet): Update tooltip styles and positioning

alisonjoseph committed:
fix: update tooltip positioning for inline snippet

asudoh reviewed (approved):
LGTM 👍 - Thanks @alisonjoseph!

tay-aitken reviewed (commented):
looks good!
